<a href="https://colab.research.google.com/github/Rashmi706/Intro-to-ML-Rashmi-K-S/blob/master/classification_redwinequality.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>




**Introduction**

This is a project where we need to compare several Classification algorithms to predict wine quality which has score between 0 and 10.We need to compare and select an algorithm to find out what makes a good wine by using winequality-red.csv.

**Contents of data file**

The attributes in this dataset are

1. fixed acidity
2. volatile acidity
3. citric acid
4. residual sugar
5. chlorides
6. free sulfur dioxide
7. total sulfur dioxide
8. density
9. pH
10. sulphates
11. alcohol
12. quality(score between 0 and 10)

**Data preprocessing**

In this step of analysis there are some of the features to train and test the machine learning model and the target to predict which is 'quality' that are defined.And then Standardization is done for the features because different scales of features may impact the performance of machine learning models.For this purpose,StandardScalar() function is used which is defined in sklearn.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
dataset = pd.read_csv('/content/winequality-red.csv')

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [ ]:
dataset['quality'].unique()

array([5, 6, 7, 4, 8, 3])

In [ ]:
X = dataset.iloc[: , 1:-1].values
y = dataset.iloc[:, -1].values

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X ,y, test_size=0.20)

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
sc = StandardScaler()

In [ ]:
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [ ]:
l_cla = LogisticRegression()
k_cla = KNeighborsClassifier()
d_cla = DecisionTreeClassifier()
r_cla = RandomForestClassifier()
s_cla = SVC(kernel='linear')
ks_cla = SVC(kernel= 'rbf')

In [ ]:
l_cla.fit(X_train, y_train)
k_cla.fit(X_train, y_train)
d_cla.fit(X_train, y_train)
r_cla.fit(X_train, y_train)
s_cla.fit(X_train, y_train)
ks_cla.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [ ]:
l_pred = l_cla.predict(X_test)
k_pred = k_cla.predict(X_test)
d_pred = d_cla.predict(X_test)
r_pred = r_cla.predict(X_test)
s_pred = s_cla.predict(X_test)
ks_pred = ks_cla.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
l_c = confusion_matrix(y_test, l_pred)
k_c = confusion_matrix(y_test, k_pred)
d_c = confusion_matrix(y_test, d_pred)
r_c = confusion_matrix(y_test, r_pred)
s_c = confusion_matrix(y_test, s_pred)
ks_c = confusion_matrix(y_test, ks_pred)

In [ ]:
l_c

array([[ 0,  1,  0,  0,  0,  0],
       [ 0,  0,  7,  6,  0,  0],
       [ 0,  0, 98, 26,  1,  0],
       [ 0,  0, 38, 84,  7,  0],
       [ 0,  0,  4, 28, 17,  0],
       [ 0,  0,  0,  1,  2,  0]])

In [ ]:
k_c

array([[ 0,  0,  1,  0,  0,  0],
       [ 0,  0,  6,  7,  0,  0],
       [ 0,  2, 87, 36,  0,  0],
       [ 0,  0, 39, 80, 10,  0],
       [ 0,  0,  4, 25, 20,  0],
       [ 0,  0,  0,  1,  2,  0]])

In [ ]:
d_c

array([[ 0,  0,  0,  1,  0,  0],
       [ 1,  3,  2,  7,  0,  0],
       [ 0,  2, 96, 22,  5,  0],
       [ 1,  3, 42, 73,  9,  1],
       [ 0,  0,  3, 17, 27,  2],
       [ 0,  0,  0,  0,  3,  0]])

In [ ]:
r_c

array([[  0,   0,   1,   0,   0,   0],
       [  1,   0,   6,   6,   0,   0],
       [  0,   0, 105,  19,   1,   0],
       [  0,   0,  27,  97,   5,   0],
       [  0,   0,   2,  23,  23,   1],
       [  0,   0,   0,   2,   1,   0]])

In [ ]:
s_c

array([[ 0,  0,  1,  0,  0,  0],
       [ 0,  0,  7,  6,  0,  0],
       [ 0,  0, 98, 27,  0,  0],
       [ 0,  0, 44, 85,  0,  0],
       [ 0,  0,  4, 45,  0,  0],
       [ 0,  0,  0,  3,  0,  0]])

In [ ]:
ks_c

array([[ 0,  0,  1,  0,  0,  0],
       [ 0,  0,  7,  6,  0,  0],
       [ 0,  0, 95, 30,  0,  0],
       [ 0,  0, 38, 88,  3,  0],
       [ 0,  0,  3, 30, 16,  0],
       [ 0,  0,  0,  2,  1,  0]])

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
l_a = accuracy_score(y_test, l_pred)
k_a = accuracy_score(y_test, k_pred)
d_a = accuracy_score(y_test, d_pred)
r_a = accuracy_score(y_test, r_pred)
s_a = accuracy_score(y_test, s_pred)
ks_a = accuracy_score(y_test, ks_pred)

In [ ]:
print('Logistic Regression: ' + str(l_a) + '\nKNN: ' + str(k_a) + '\nDecision Tree: ' + str(d_a) + '\nRandom Forest: ' + str(r_a) + '\nLinear SVC: ' + str(s_a) + '\nKernel SVC: ' + str(l_a))

Logistic Regression: 0.621875
KNN: 0.584375
Decision Tree: 0.621875
Random Forest: 0.703125
Linear SVC: 0.571875
Kernel SVC: 0.621875


**Algorithms tried on data**

1. Logistic Regression
2. K-Nearest Neighbors Classification
3. Decision Tree Classification
4. Random Forest Classification
5. Linear Support Vector Classification
6. Kernel Support Vector Classification

**Results**

According to the given dataset,the following results are obtained

Logistic Regression: 0.621875

KNN: 0.584375

Decision Tree: 0.621875

Random Forest: 0.703125

Linear SVC: 0.571875

Kernel SVC: 0.621875

**Conclusion**

Based on the results we come to a conclusion that not all input features are essential and effect the data.For classifying the wine quality, we have implemented multiple algorithms.We were able to achieve maximum accuracy using Random Forest Classification algorithm.